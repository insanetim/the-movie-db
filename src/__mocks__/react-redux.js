export const dispatch = jest.fn()

export const useDispatch = jest.fn(() => dispatch)
export const useSelector = jest.fn(fn => fn())
