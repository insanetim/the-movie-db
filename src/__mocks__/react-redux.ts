const dispatch = jest.fn()
const getState = jest.fn()

const useDispatch = jest.fn(() => dispatch)
const useSelector = jest.fn(fn => fn())

export { dispatch, getState, useDispatch, useSelector }
