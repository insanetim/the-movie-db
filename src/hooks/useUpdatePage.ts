import { AsyncThunkAction, Dispatch } from '@reduxjs/toolkit'
import { SetURLSearchParams } from 'react-router-dom'
import getParams from 'src/utils/helpers/getParams'

import { useAppDispatch } from './useRedux'

declare type AsyncThunkConfig = {
  dispatch?: Dispatch
  extra?: unknown
  fulfilledMeta?: unknown
  pendingMeta?: unknown
  rejectValue?: unknown
  rejectedMeta?: unknown
  serializedErrorType?: unknown
  state?: unknown
}

export type UseUpdatePageProps = {
  action: AsyncThunkAction<unknown, unknown, AsyncThunkConfig>
  items: undefined | unknown[]
  page: string
  setSearchParams: SetURLSearchParams
}

const useUpdatePage = ({
  action,
  items: itemsProp,
  page: pageProp,
  setSearchParams: setSearchParamsProp,
}: UseUpdatePageProps) => {
  const dispatch = useAppDispatch()

  const updatePage = () => {
    const items = itemsProp || []
    const page = Number(pageProp)

    if (items.length === 1 && page > 1) {
      setSearchParamsProp(getParams({ page: page - 1 }))
    } else {
      dispatch(action)
    }
  }

  return { updatePage }
}

export default useUpdatePage
