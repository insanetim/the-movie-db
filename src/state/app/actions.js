import { v4 as uuid } from 'uuid'

import { NOTIFICATION_DURATION, NOTIFICATION_TYPE } from 'src/constants'

import * as types from './types'

export const showModal = payload => ({
  type: types.SHOW_MODAL,
  payload
})

export const hideModal = () => ({
  type: types.HIDE_MODAL
})

export const showNotification = ({
  id = uuid(),
  messageType = NOTIFICATION_TYPE.SUCCESS,
  messageText,
  duration = NOTIFICATION_DURATION
}) => ({
  type: types.SHOW_NOTIFICATION,
  payload: {
    id,
    messageType,
    messageText,
    duration
  }
})

export const hideNotification = payload => ({
  type: types.HIDE_NOTIFICATION,
  payload
})
