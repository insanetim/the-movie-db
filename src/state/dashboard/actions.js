import * as types from './types'

export const fetchTrending = page => ({
  type: types.FETCH_TRENDING,
  payload: page
})

export const fetchTrendingRequest = () => ({
  type: types.FETCH_TRENDING_REQUEST
})

export const fetchTrendingSuccess = movies => ({
  type: types.FETCH_TRENDING_SUCCESS,
  payload: movies
})

export const fetchTrendingError = error => ({
  type: types.FETCH_TRENDING_ERROR,
  payload: error
})

export const setTrendingPage = page => ({
  type: types.SET_TRENDING_PAGE,
  payload: page
})

export const fetchSearch = ({ page, query }) => ({
  type: types.FETCH_SEARCH,
  payload: { page, query }
})

export const fetchSearchRequest = () => ({
  type: types.FETCH_SEARCH_REQUEST
})

export const fetchSearchSuccess = movies => ({
  type: types.FETCH_SEARCH_SUCCESS,
  payload: movies
})

export const fetchSearchError = error => ({
  type: types.FETCH_SEARCH_ERROR,
  payload: error
})

export const setSearchPage = page => ({
  type: types.SET_SEARCH_PAGE,
  payload: page
})
