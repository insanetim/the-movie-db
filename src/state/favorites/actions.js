import * as types from './types'

export const fetchFavorites = page => ({
  type: types.FETCH_FAVORITES,
  payload: page
})

export const fetchFavoritesRequest = page => ({
  type: types.FETCH_FAVORITES_REQUEST,
  payload: page
})

export const fetchFavoritesSuccess = movies => ({
  type: types.FETCH_FAVORITES_SUCCESS,
  payload: movies
})

export const fetchFavoritesFailure = error => ({
  type: types.FETCH_FAVORITES_FAILURE,
  payload: error
})

export const setFavoritesPage = page => ({
  type: types.SET_FAVORITES_PAGE,
  payload: page
})
