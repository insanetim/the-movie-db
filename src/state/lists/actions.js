import * as types from './types'

export const fetchLists = ({ page, callback }) => ({
  type: types.FETCH_LISTS,
  payload: { page, callback }
})

export const fetchListsRequest = page => ({
  type: types.FETCH_LISTS_REQUEST,
  payload: page
})

export const fetchListsSuccess = movies => ({
  type: types.FETCH_LISTS_SUCCESS,
  payload: movies
})

export const fetchListsFailure = error => ({
  type: types.FETCH_LISTS_FAILURE,
  payload: error
})

export const setListsPage = page => ({
  type: types.SET_LISTS_PAGE,
  payload: page
})

export const fetchList = listId => ({
  type: types.FETCH_LIST,
  payload: listId
})

export const fetchListRequest = () => ({
  type: types.FETCH_LIST_REQUEST
})

export const fetchListSuccess = movies => ({
  type: types.FETCH_LIST_SUCCESS,
  payload: movies
})

export const fetchListFailure = error => ({
  type: types.FETCH_LIST_FAILURE,
  payload: error
})

export const createList = ({ value, callback }) => ({
  type: types.CREATE_LIST,
  payload: { value, callback }
})

export const addToList = ({ listId, movieId }) => ({
  type: types.ADD_TO_LIST,
  payload: { listId, movieId }
})

export const removeFromList = ({ listId, movieId }) => ({
  type: types.REMOVE_FROM_LIST,
  payload: { listId, movieId }
})

export const deleteList = ({ listId, callback }) => ({
  type: types.DELETE_LIST,
  payload: { listId, callback }
})
