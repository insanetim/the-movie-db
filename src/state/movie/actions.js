import * as types from './types'

export const fetchMovie = movieId => ({
  type: types.FETCH_MOVIE,
  payload: movieId
})

export const fetchMovieRequest = () => ({
  type: types.FETCH_MOVIE_REQUEST
})

export const fetchMovieSuccess = movie => ({
  type: types.FETCH_MOVIE_SUCCESS,
  payload: movie
})

export const fetchMovieFailure = error => ({
  type: types.FETCH_MOVIE_FAILURE,
  payload: error
})

export const fetchMovieStates = movieId => ({
  type: types.FETCH_MOVIE_STATES,
  payload: movieId
})

export const updateMovieStates = movieStates => ({
  type: types.UPDATE_MOVIE_STATES,
  payload: movieStates
})

export const changeMovieInFavorites = ({ movieId, inFavorites }) => ({
  type: types.CHANGE_MOVIE_IN_FAVORITES,
  payload: { movieId, inFavorites }
})

export const changeMovieInWatchlist = ({ movieId, inWatchlist }) => ({
  type: types.CHANGE_MOVIE_IN_WATCHLIST,
  payload: { movieId, inWatchlist }
})
