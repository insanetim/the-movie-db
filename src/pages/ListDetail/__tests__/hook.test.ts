import { act, renderHook } from '@testing-library/react'
import { Modal } from 'antd'
import { useNavigate, useParams, useSearchParams } from 'react-router-dom'
import { dispatch } from 'src/__mocks__/react-redux'
import useUpdatePage from 'src/hooks/useUpdatePage'
import * as listsActions from 'src/store/lists/actions'

import useContainer from '../hook'

jest.mock('src/store/lists/selectors', () => ({
  listDetailErrorSelector: () => null,
  listDetailLoadingSelector: () => true,
  listDetailSelector: () => null
}))

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
  useParams: jest.fn(() => ({ listId: '123' })),
  useSearchParams: jest.fn()
}))
const navigate = jest.fn()
jest.mocked(useNavigate).mockReturnValue(navigate)
const searchParams = new URLSearchParams()
const setSearchParams = jest.fn()
jest.mocked(useSearchParams).mockReturnValue([searchParams, setSearchParams])

jest.mock('src/hooks/useUpdatePage')
const updatePage = jest.fn()
jest.mocked(useUpdatePage).mockReturnValue({ updatePage })

describe('ListDetail useContainer hook', () => {
  const modalSpy = jest.spyOn(Modal, 'confirm')

  it('should match snapshot', () => {
    const { result } = renderHook(useContainer)

    expect(result.current).toMatchSnapshot()
  })

  it('should match snapshot without params', () => {
    jest.mocked(useParams).mockReturnValueOnce({})
    const { result } = renderHook(useContainer)

    expect(result.current).toMatchSnapshot()
  })

  it('should check "handlePagination" method', () => {
    const { result } = renderHook(useContainer)

    act(() => {
      result.current.handlePagination(3)
    })

    expect(setSearchParams).toHaveBeenCalledWith({ page: '3' })
  })

  it('should check "handleListDelete" method', async () => {
    const deleteList = jest.spyOn(listsActions, 'deleteList')
    let onOk = () => {}
    const { result } = renderHook(useContainer)

    await act(() => {
      onOk = result.current.handleListDelete()
      onOk()
    })

    expect(modalSpy).toHaveBeenCalledWith({
      onOk,
      title: 'Do you want to delete list?'
    })
    expect(dispatch).toHaveBeenCalled()
    expect(deleteList).toHaveBeenCalledWith('123')
    expect(navigate).toHaveBeenCalledWith('/lists')
  })

  it('should check "handleMovieDelete" method', async () => {
    const removeFromList = jest.spyOn(listsActions, 'removeFromList')
    let onOk = () => {}
    const { result } = renderHook(useContainer)

    await act(() => {
      onOk = result.current.handleMovieDelete(123, {
        stopPropagation: jest.fn()
      } as never)
      onOk()
    })

    expect(modalSpy).toHaveBeenCalledWith({
      onOk,
      title: 'Do you want to delete movie from this list?'
    })
    expect(dispatch).toHaveBeenCalled()
    expect(removeFromList).toHaveBeenCalledWith({ listId: '123', movieId: 123 })
    expect(updatePage).toHaveBeenCalled()
  })

  it('should check "useEffect" method', () => {
    const fetchListDetail = jest.spyOn(listsActions, 'fetchListDetail')
    renderHook(useContainer)

    expect(dispatch).toHaveBeenCalled()
    expect(fetchListDetail).toHaveBeenCalledWith({ listId: '123', page: '1' })
  })
})
