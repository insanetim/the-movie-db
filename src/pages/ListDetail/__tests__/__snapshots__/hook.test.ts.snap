// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListDetail useContainer hook should match snapshot 1`] = `
{
  "error": null,
  "handleListDelete": [Function],
  "handleMovieDelete": [Function],
  "handlePagination": [Function],
  "list": null,
  "loading": true,
}
`;

exports[`ListDetail useContainer hook should match snapshot without params 1`] = `
{
  "error": null,
  "handleListDelete": [Function],
  "handleMovieDelete": [Function],
  "handlePagination": [Function],
  "list": null,
  "loading": true,
}
`;
